File: tests/csharp/modern_features/null_conditional_operator.cs
// vybe-test: csharp/modern_features/null_conditional_operator
// origin: languages/csharp/tests/csharp/test_modern_features.rs

using static __Harness;

string s = null;
__P((s?.ToUpper() ?? "null").ToString());
s = "hello";
__P((s?.ToUpper() ?? "null").ToString());
__Check("null\nHELLO");

public static class __Harness {
    public static string __buf = "";
    public static void __P(string s) { __buf = __buf + s + "\n"; }
    public static void __Pr(string s) { __buf = __buf + s; }
    public static void __Check(string want) {
        if (__buf != want && __buf != want + "\n") {
            Console.WriteLine("FAIL: want [" + want + "] got [" + __buf + "]");
            throw new Exception("assertion failed");
        }
    }
}

File: tests/vb/vb_null_conditional/null_conditional_operator.vb
' vybe-test: vb/vb_null_conditional/null_conditional_operator
' origin: languages/vb/tests/vb/test_vb_null_conditional.rs

' Vybe test harness — Visual Basic.
'
' Real VB source alongside harness/go/check.go and harness/js/check.js, the way
' test262's assert.js is JavaScript.
'
' A test's verdict is its EXIT CODE. __Check prints its diagnostic BEFORE
' throwing: an uncaught exception surfaces as `RuntimeError: [object]`, which
' says nothing at all.

Module VybeCheck
    Sub __Check(got As String, want As String)
        If got <> want Then
            Console.WriteLine("FAIL: want [" & want & "] got [" & got & "]")
            Throw New Exception("assertion failed")
        End If
    End Sub
End Module

Class Data
    Public Property Value As String
End Class

Module M
    Sub Main()
        Dim d As Data = Nothing
        
        ' Using ? before dot checks if d is nothing
        Dim len As Integer? = d?.Value?.Length
        
        __Check(CStr(len.HasValue), "False")
        
        d = New Data() With { .Value = "Test" }
        len = d?.Value?.Length
        __Check(CStr(len.HasValue), "True")
        __Check(CStr(len.Value), "4")
    End Sub
End Module
